cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -402,30 +420,83 @@ public boolean apply(RunRecordMeta input) {
     MDSKey activeKey = getProgramKeyBuilder(recordType, programId).build();
 
-    return list(activeKey, null, RunRecordMeta.class, limit, andPredicate(new Predicate<RunRecordMeta>() {
-                  @Override
-                  public boolean apply(RunRecordMeta input) {
-                    return input.getStartTs() >= startTime && input.getStartTs() < endTime;
-                  }
-                }, filter));
+    Map<MDSKey, RunRecordMeta> runRecordMap = listKV(activeKey, null, RunRecordMeta.class, limit,
+                                                     andPredicate(new Predicate<RunRecordMeta>() {
+      @Override
+      public boolean apply(RunRecordMeta input) {
+        return input.getStartTs() >= startTime && input.getStartTs() < endTime;
+      }
+    }, filter));
+
+    return (getProgramRunIdMap(runRecordMap));
+  }
+
+  private  Map<ProgramRunId, RunRecordMeta> getSuspendedRuns(HashSet<ProgramRunId> programRunIds, long startTime,
+                                                             long endTime, int limit,
+                                                             @Nullable Predicate<RunRecordMeta> filter) {
+    return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_SUSPENDED, startTime, endTime, limit, filter);
+  }
+
+  private  Map<ProgramRunId, RunRecordMeta> getActiveRuns(HashSet<ProgramRunId> programRunIds, long startTime,
+                                                          long endTime, int limit,
+                                                          @Nullable Predicate<RunRecordMeta> filter) {
+    return getNonCompleteRuns(programRunIds, TYPE_RUN_RECORD_STARTED, startTime, endTime, limit, filter);
+  }
+
+  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(HashSet<ProgramRunId> programRunIds, String recordType,
+                                                              long startTime, long endTime, int limit,
+                                                              @Nullable Predicate<RunRecordMeta> filter) {
+    MDSKey activeKey = getProgramKeyBuilder(recordType, null).build();
+    MDSKey start = new MDSKey.Builder(activeKey).add(getInvertedTsScanKeyPart(endTime)).build();
+    MDSKey stop = new MDSKey.Builder(activeKey).add(getInvertedTsScanKeyPart(startTime)).build();
+
+    return getProgramRunIdMap(listKV(start, stop, RunRecordMeta.class, limit, programRunIds,
+                  filter == null ? Predicates.<RunRecordMeta>alwaysTrue() : filter));
+  }
+
+  private  Map<MDSKey, RunRecordMeta> getHistoricalRuns(HashSet<ProgramRunId> programRunIds,
+                                                        final long startTime, final long endTime, int limit,
+                                                        @Nullable Predicate<RunRecordMeta> filter) {
+    MDSKey historyKey = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, null).build();
+
+    MDSKey start = new MDSKey.Builder(historyKey).add(getInvertedTsScanKeyPart(endTime)).build();
```

Comment:
Added optimization with  new implementation of listKV which builds a fuzzyrow filter out of all the keys built from all ProgramIDs.